doc: add document for testing with seastar

tchaikov opened the pull request:
since Boost.Test does not provide builtin support for testing test cases implemented with coroutine. we have our own test runner and macros for declaring tests. for new developers these macros and the macros like `BOOST_AUTO_TEST_CASE` could be confusing.

this document intend to clarify the declaration and organization of tests in Seastar and Seastar applications using Seastar's testing framework.

lersek commented on doc/testing.md:
```diff
@@ -0,0 +1,51 @@
+
+* Boost.Test Native: Tests defined using `BOOST_AUTO_TEST_CASE` and related macros from Boost.Test. For more information, see the [Boost Test documentation](https://www.boost.org/doc/libs/release/libs/test/doc/html/boost_test/utf_reference/test_org_reference.html).
+* Coroutine: Tests defined using `SEASTAR_TEST_CASE`. The test body returns a future, allowing implementation as a coroutine.
+* Coroutine with `seastar::thread`: Tests defined using `SEASTAR_THREAD_TEST_CASE`. The test body is launched in a Seastar thread, allowing the use of Seastar coroutines. These tests should return `void`.
```
The link to the Boost.Test documentation is helpful; can we do the same with the other two kinds of tests?
- https://github.com/scylladb/seastar/blob/master/doc/tutorial.md#coroutines
- https://github.com/scylladb/seastar/blob/master/doc/tutorial.md#seastarthread

tchaikov replied:
sure. added the hyper links.

lersek commented on doc/testing.md:
```diff
@@ -0,0 +1,51 @@
+* For Seastar facilities that don't depend on coroutines, consider using `BOOST_AUTO_TEST_CASE`.
+
+
+## Tests organization
```
`Tests' Organization` or `Organizing Tests` look more correct IMO.

tchaikov replied:
i picked the former.

lersek commented on doc/testing.md:
```diff
@@ -0,0 +1,51 @@
+
+## Tests organization
+
+* Tests defined with `BOOST_AUTO_TEST_CASE` are driven by Boost.Test's built-in test runner. They require defining `BOOST_TEST_MODULE` to include the runner in the executable. These tests support additional features like fixtures and data-driven testing.
```
Might want to mention that the *replacement list* for the `BOOST_TEST_MODULE` macro is supposed to be the "name of the master test suite".

tchaikov replied:
i am adding a hyper link to Boost.Test's document. it should do a better job than i do in this regard.

tchaikov commented on doc/testing.md:
```diff
@@ -0,0 +1,51 @@
+
+Seastar leverages Boost.Test and provides facilitiets for developers to implement tests in coroutines.
+
+## Tests Declaration
```
i'd go with "Test Declarations" to be more consistent with the titles of the same level.

tchaikov commented:
v2:

- fixed the misspelling of "facilitiets"
- s/Tests Declaration/Test Declarations/
- added hyperlinks for "coroutine" and "seastar::thread"
- added hyperlink for "BOOST_TEST_MODULE"

lersek reviewed (approved):
Thanks for putting up with my hair-splitting; looks great!

tchaikov commented on doc/testing.md:
```diff
@@ -0,0 +1,51 @@
+# Testing
+
+Seastar leverages Boost.Test and provides facilitiets for developers to implement tests in coroutines.
```
yeah, Benny suggested the similar before. =( i will try M-x flyspell-buffer next time.
